### PR TITLE
gui: Remove WalletView and BitcoinGUI circular dependency

### PR DESCRIPTION
**Pull request by promag**
Essentially moves the code in `WalletView::setBitcoinGUI` to the only caller. Two new signals are added beforehand in the first commit so that the connections in `WalletFrame` are all from the wallet view.

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK.

File: src/qt/walletframe.cpp [hunk 1]
```diff
@@ -62,6 +61,14 @@ bool WalletFrame::addWallet(WalletModel *walletModel)
     mapWalletViews[walletModel] = walletView;
 
     connect(walletView, &WalletView::outOfSyncWarningClicked, this, &WalletFrame::outOfSyncWarningClicked);
+    connect(walletView, &WalletView::transactionClicked, gui, &BitcoinGUI::gotoHistoryPage);
+    connect(walletView, &WalletView::coinsSent, gui, &BitcoinGUI::gotoHistoryPage);
+    connect(walletView, &WalletView::message, [this](const QString& title, const QString& message, unsigned int style) {
+        gui->message(title, message, style);
+    });
```
Comment (hebasto) [inline src/qt/walletframe.cpp]: s/`[this]`/`[gui]`/ ?

**Review by hebasto — CHANGES_REQUESTED** [review]

File: src/qt/walletframe.cpp (hunk 1 above)
Comment (promag) [reply src/qt/walletframe.cpp]: Doesn't work, `gui` is a member.

Comment (hebasto) [reply src/qt/walletframe.cpp]: oops

`BitcoinGUI::message` is a slot, and there is no need to capture local variable `gui`.
Why not use non-lambda connection syntax?

Comment (promag) [reply src/qt/walletframe.cpp]: Right, originally couldn't connect to slot because of that. I think we can "clean" these later.

**Comment by Empact** [comment]
Concept ACK

**Comment by hebasto** [comment]
w.r.t. https://github.com/bitcoin/bitcoin/pull/17500#issuecomment-575088028

It seems a bit of formalization could be helpful for our community. How about:

_**Qt Slots and Signals**
Do no make connections to external signals or slots in the `QObject` subclass implementation.
Rationale: Do not introduce needless dependencies._

**Comment by promag** [comment]
@hebasto looks good!

**Comment by promag** [comment]
Rebased and reverted to previous version since #17965 was merged.

**Comment by promag** [comment]
Appveyor status is wrong, it actually passed.

**Review by hebasto — APPROVED** [review]
ACK cb8a86d9f952401eaad68b2e3818ce50f7befd91, tested on Linux Mint 19.3.

**Review by jonasschnelli — APPROVED** [review]
utACK cb8a86d9f952401eaad68b2e3818ce50f7befd91

**Pull request merged**